pkg/types/chain_reader.go

```diff
@@ -0,0 +1,64 @@
+package types
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type chainReaderError string
+
+func (e chainReaderError) Error() string { return string(e) }
+
+const (
+	ErrInvalidType   = chainReaderError("invalid type")
+	ErrInvalidConfig = chainReaderError("invalid configuration")
+)
+
+func UnwrapClientError(err error) error {
+	if s, ok := status.FromError(err); ok {
+		if s.Code() == codes.Unimplemented {
+			return fmt.Errorf("%w : %s", errors.ErrUnsupported, s.String())
```

Comment:
I'm not sure it is correct to assume that all `Unimplemented` errors should be translated to `ErrUnsupported`. It may make perfect sense in our cases to map them 1-1, but both of these errors can come from other sources.

Reply:
Yeah, both good points.

The advantage of using the `Unimplemented` status code is that we can detect the type of error easily, but also include more detail on it in the error string.  Functionally, that would work better.  But as you say, the normal meaning of `Unimplemented` seems a bit off from what we're trying to use it for and it could be confusing.  So overall I guess it's probably better to go with `ErrUnsupported`.

The only disappointing thing about `ErrUnsupported` is that it forces us to choose between being able to detect unsupported errors with `errors.Is(err, ErrUnsupported)` and being able to include more detail than just "unsupported".  I guess omitting the detail is okay though.